### ISSUE TITLE
Fix causes duplicate table aliases by simply removing them when table appears twice in join

### ISSUE THREAD

**Issue by OJFord** [sqlfluff/sqlfluff#610]
> L020 | Duplicate table alias

Since after applying `sqlfluff fix` on for example:
```sql
select *
from a a1
join b on a1.i = b.j
join a a2 on a2.i = b.k
```
becomes:
```sql
select *
from a
join b on a.i = b.j
join a on a.i = b.k
```

Workaround I think is to move part of the join to a `with` CTE, e.g. in this case:
```sql
with a2 as (select i from a)
select *
from a
join b on a.i = b.j
join a2 on a2.b.k
```

**Comment by barrywhart** [comment]
I tested with:
* the latest code in `master` as of [today](https://github.com/sqlfluff/sqlfluff/pull/596/commits/ce673c5c7731c3c7e0fff76f77d49bc8e362d7f2) 
* the code as of [December 3](https://github.com/sqlfluff/sqlfluff/pull/596/commits/2b7bc9db611cd5df893826f98fad61d289081d8a)

I ran `sqlfluff fix test.sql`. In both cases, the resulting SQL file was the same, and it didn't remove the alias on the second join with table `a`.

```
select *
from a
join b on a.i = b.j
join a as a2 on a2.i = b.k
```

I also tested with version 0.3.6, running the command `sqlfluff fix --rules L020 test.sql`. In this case, `sqlfluff fix` reported `no fixable linting violations found` and did not change the SQL file.

@OJFord, can you clarify what version of SQLFluff you were using when you saw this behavior? It may also be helpful if you can provide the specific command you ran and the contents of your `.sqlfluff` file. Thanks!

**Comment by OJFord** [comment]
It was master at the time (via `rev: master` in the pre-commit hook). No `.sqlfluff`, `tox.ini`:
```ini
[sqlfluff]
dialect = postgres
```

It was just a first pass to try to add SQL linting to an existing project, but unfortunately hit quite a few errors, this seemed to be most common from what I recall. I'll hopefully have another stab at it soon.

**Comment by OJFord** [comment]
I just tried again with v0.4.1 - what's odd to me is that fix succeeds (after four iterations before no diff) but lint fails, even on the same files... How can it fail to parse - or even lex in some cases - and yet the fixer can modify it fine (the output looks sensible)?

`COMMENT ON VIEW name IS 'comment';` is one that fails 'PRS' with `sqlfluff-lint`, but the file is reformatted no problem.

That, `$$`, and `CREATE OR REPLACE FUNCTION` seem to be the bulk of the errors.

Anyway, that's beside this issue, I still see the issue described though - but perhaps I got a crucial detail wrong. Where it's actually happening (at least the case I'm looking at now) is two `join`s, rather than a `from` and a `join`.

Simple reproduction:
```sql
select *
from aaaaaa a
join bbbbbb b on b.a = a.id
join bbbbbb b2 on b2.other = b.id
```

becomes:
```sql
select *
from aaaaaa
join bbbbbb on bbbbbb.a = aaaaaa.id
join bbbbbb on bbbbbb.other = bbbbbb.id
```

**Comment by barrywhart** [comment]
About the parsing errors, when SQLFluff parser is not an "all or nothing" thing. It generates a partial "best effort" parse tree even when it can't parse everything. Some but not all of the rules/fixes are configured to run when the parse fails. Generally, these are the simpler rules that look at small parts of the SQL.

**Comment by barrywhart** [comment]
Thanks for the updated report. Knowing the version should help us reproduce and fix this.
